src/ManifoldMuseum.jl

```diff
@@ -102,29 +103,105 @@ Keyword arguments can be used to specify tolerances.
 isapprox(M::Manifold, x, v, w; kwargs...) = isapprox(v, w; kwargs...)
 
 """
-    retract!(M::Manifold, y, x, v, t=1)
+    OutOfInjectivityRadiusError
+
+An error thrown when a function (for example logarithmic map or inverse
+retraction) is given arguments outside of its injectivity radius.
+"""
+struct OutOfInjectivityRadiusError <: Exception end
+
+abstract type AbstractRetractionMethod end
+
+"""
+    ExponentialRetraction
+
+Retraction using the exponential map.
+"""
+struct ExponentialRetraction <: AbstractRetractionMethod end
+
+"""
+    retract!(M::Manifold, y, x, v, [t=1], [method::AbstractRetractionMethod=ExponentialRetraction()])
 
 Retraction (cheaper, approximate version of exponential map) of tangent
 vector `t*v` at point `x` from manifold `M`.
 Result is saved to `y`.
+
+Retraction method can be specified by the last argument. Please look at the
+documentation of respective manifolds for available methods.
 """
-retract!(M::Manifold, y, x, v) = exp!(M, y, x, v)
+retract!(M::Manifold, y, x, v, method::ExponentialRetraction) = exp!(M, y, x, v)
+
+retract!(M::Manifold, y, x, v) = retract!(M, y, x, v, ExponentialRetraction())
+
+retract!(M::Manifold, y, x, v, t::Number) = retract!(M, y, x, t*v)
```

Comment:
Thanks, I didn't notice that Distributions has that function (it's not in the docs).

The problem with `t` being real is that `Dual` numbers (from e.g. ForwardDiff) derive from `Number` but not `Real`, so AD will fail here.

Reply:
> The problem with `t` being real is that `Dual` numbers (from e.g. ForwardDiff) derive from `Number` but not `Real`, so AD will fail here.

`Dual` from `ForwardDiff` at least is derived from `Real`. https://github.com/JuliaDiff/ForwardDiff.jl/blob/master/src/dual.jl#L14 (and [here](http://www.juliadiff.org/ForwardDiff.jl/stable/user/limitations.html) the documentation recommends `Real` as generic enough). Are there other packages this will cause a problem for?

Reply:
Thanks, I must've mistaken it with something else. `Dual` from DualNumbers is not `Real` though: https://github.com/JuliaDiff/DualNumbers.jl/blob/677e2d52bfadc365f2b3f43b51c1dc66d284a9fa/src/dual.jl#L3 . There is likely a lot of other places where DualNumbers won't work for some manifolds so I'll make this `Real`.